sdk/identity/azure_identity/src/credentials/workload_identity_credentials.rs

```diff
@@ -155,3 +152,60 @@ impl TokenCredential for WorkloadIdentityCredential {
         self.cache.clear().await
     }
 }
+
+#[derive(Debug)]
+enum Token {
+    Value(Secret),
+    File {
+        path: String,
+        cache: Arc<RwLock<FileCache>>,
+    },
+}
+
+#[derive(Debug)]
+struct FileCache {
+    token: Secret,
+    last_read: Instant,
+}
+
+impl Token {
+    fn with_file(path: &str) -> azure_core::Result<Self> {
+        let last_read = Instant::now();
+        let token = std::fs::read_to_string(path).with_context(ErrorKind::Credential, || {
+            format!("failed to read federated token from file {}", path)
+        })?;
+
+        Ok(Self::File {
+            path: path.into(),
+            cache: Arc::new(RwLock::new(FileCache {
+                token: Secret::new(token),
+                last_read,
+            })),
+        })
+    }
+
+    async fn secret(&self) -> azure_core::Result<String> {
+        match self {
+            Self::Value(secret) => Ok(secret.secret().into()),
+            Self::File { path, cache } => {
+                const TIMEOUT: Duration = Duration::from_secs(600);
+
+                let now = Instant::now();
+                let cache = cache.upgradable_read().await;
+                if now - cache.last_read > TIMEOUT {
+                    let token = std::fs::read_to_string(path)
```

Comment:
This would in theory block the executor

Reply:
Same goes here: https://github.com/Azure/azure-sdk-for-rust/blob/6d89f3acee27d7c26bca238de36fe91416690c1d/sdk/identity/azure_identity/src/credentials/client_certificate_credentials.rs#L311

I suppose the most fitting requirements of being executor agnostic would be if there was a trait for all sorts of I/O, including launching processes (#1652) which then has a default implementation via tokio, but also can have implementations via other executors (or even blocking ones).

But let's not have perfect be the enemy of good. I'd say it's an orthogonal concern.

Reply:
I did consider this and was looking for existing abstractions that don't require a runtime - not finding any, which isn't actually surprising. It'd be great if more abstractions where in/under `std::future` so that we, the library developers, aren't always having to define our own incompatible (with other libs) abstractions, but it's likely what we'll have to do.

I've created #2002 but would add that this is no worse than what the code was doing before. Not ideal, I agree, but we'll need to create abstractions not just for the HTTP stack (and AMQP, as we've done), but processes as noted in #1652 and file system as noted in #2002 now.

Reply:
We do already depend on the `futures` crate so that should be fine in the short term.